### PR TITLE
Add TensorMechanics automatic differentiation example

### PR DESCRIPTION
**Pull request by lindsayad**
- Template RankTwoTensor, RankFourTensor, ColumnMajorMatrix
- Dynamic memory for `MooseADWrapper._dual_number`
    - This limits memory usage by non-AD applications while maintaining inter-operability of AD and non-AD objects for any simulations that are using AD

### CODE REVIEW

**Comment by tophmatthews** [comment]
I would give multiple hurray party hats if I could! (I can't, I checked)

**Comment by lindsayad** [comment]
I'll note that changing the implementation of the `MooseADWrapper::_dual_number` from stack to heap allocated has a virtually non-existent impact on speed. I've profiled our `ad_coupled_material` test with `-r 7 ADKernels/diff/prop_to_use='AdReg'` which forces heap allocation of the `DualNumber` in the `MooseADWrapper::dn` method. The results are:

Stack:

screen shot 2018-12-12 at 11 30 46 am

Heap:

screen shot 2018-12-12 at 11 30 11 am

I would contend that a cumulative time of .18% is well worth:

1. the memory savings this should provide applications not using AD, especially those using larger mathematical structures like tensors, and especially those using stateful materials!
2. Being able to put the MetaPhysicL includes in the `MooseADWrapper.C` file instead of the header. This is huge because if the `metaphysicl` includes are in the header, in any translation unit that pulls in `MooseADWrapper.h`, there **must** be an inclusion of `metaphysicl/dual_number_decl.h` before any inclusion of `libmesh/type_vector.h` because the `TypeVector<ADReal>::norm` function signature requires awareness of functions declared in the MetaPhysicL headers

**Comment by lindsayad** [comment]
Of course this whole `MooseADWrapper` class can be removed if we decide we no longer need inter-operability between ADKernels/Materials and Kernels/ADMaterials

File: framework/src/utils/MooseADWrapper.C [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+RankTwoTensorTempl<ADReal> &
+MooseADWrapper<RankTwoTensorTempl<Real>>::dn(bool)
+{
+  return *_dual_number;
+}
+
+void
+MooseADWrapper<RankTwoTensorTempl<Real>>::copyDualNumberToValue()
+{
+  for (decltype(LIBMESH_DIM) i = 0; i < LIBMESH_DIM; ++i)
```
Comment (dschwen) [inline framework/src/utils/MooseADWrapper.C]: decltype on a preprocessor symbol?

File: framework/src/utils/BicubicInterpolation.C [hunk 2]
```diff
@@ -75,7 +75,7 @@ BicubicInterpolation::sampleDerivative(Real x1, Real x2, unsigned int deriv_var)
 
     Real d = _x1[x1u] - _x1[x1l];
 
-    if (!MooseUtils::absoluteFuzzyEqual(d, 0.0))
+    if (!MooseUtils::absoluteFuzzyEqual(d, Real(0.0)))
```
Comment (dschwen) [inline framework/src/utils/BicubicInterpolation.C]: `Real(0.0)` id probably necessary as ADReal can be constructed by a floating point number as well. But this is clunky.

**Review by dschwen — COMMENTED** [review]
I'd be interested to see the performance impact on mechanics problems that do not use AD. Obviously this is a freight train at full steam ahead that cannot be stopped now, but I have to say the impact on code readability is high. The memory overhead for stateful material properties is not negligible, the switching of stateful properties is not non-trivial. etc.

File: framework/include/utils/MooseTypes.h [hunk 3]
```diff
@@ -77,13 +72,33 @@
...
 template <typename>
 class MaterialProperty;
 template <typename>
 class ADMaterialPropertyObject;
 
+namespace libMesh
+{
+template <typename>
+class VectorValue;
+typedef VectorValue<Real> RealVectorValue;
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/utils/MooseTypes.h]: We don't include `libmesh/vector_value.h` in `MooseTypes` (at least after this PR). I'm trying to clean-up our header dependencies a little bit.

File: framework/src/utils/BicubicInterpolation.C (hunk 2 above)
Comment (lindsayad) [reply framework/src/utils/BicubicInterpolation.C]: It looks like this cast may be unnecessary after all...

File: framework/include/utils/MooseTypes.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/utils/MooseTypes.h]: Resolving as long as that's ok with you...

**Comment by YaqiWang** [comment]
I am curious if it is possible to merge ADKernel and Kernel into one. If performance is a concern, we can use defines to turn AD on and off like what we did for AMR.

**Comment by lindsayad** [comment]
I just don't see that working. You want to combine them to reduce code duplication? That's a worthy goal except: 

- unless you want to compute derivatives both during Jacobian and residual, the `ADKernel` class must have the `ComputeStage` template parameter. Because of this there are two kernels created each time an ADKernel is requested: one for computing the residual and one for computing the Jacobian. With rattlesnake already having tons of kernels, would you really want to double the number? I would not want to try to modify this behavior on an application specific basis by using defines (e.g. if `MOOSE_ENABLE_AD` then have a class template, else have an ordinary class)
- how do you determine whether to call the `computeQpResidual` vs `computeQpJacobian` (automatic differentiation vs. hand-coded) routines when computing the Jacobian? What about the fact that we only need to call `computeQpResidual` does all the derivative calculations for every coupled variable in one pass with AD while we have to call `computeQpOffDiagJacobian` for each coupled variable when not using AD?

I think having them separate is the least disruptive and also the most functional.

**Review by dschwen — APPROVED** [review]
Looks good to me. The performance impact (CPU/Mem) is not severe (in fact it improves on the AD that is currently in HEAD).

**Pull request merged**